packages/apollo-collaboration-server/src/features/features.service.ts

```diff
@@ -0,0 +1,215 @@
+          if (
+            val.hasOwnProperty('child_features') &&
+            Object.keys(assignedVal.child_features).length > 0
+          ) {
+            // Let's add featureId to each child recursively
+            this.setAndGetFeatureIdRecursively(assignedVal, featureIdArray)
+          }
+        }
+        this.logger.verbose(
+          `So far apollo ids are: ${featureIdArray.toString()}\n`,
+        )
+      }
+      // ******* TODO : CURRENTLY WE ADD ONLY FEATURES (I.E. NOT COMMENTS, DIRECTIVES AND SEQUENCES) INTO DATABASE ****************
+      // If we are adding feature then we need to retrieve proper refSeqId
+      if (isFeature) {
+        const refSeqDoc = await this.refSeqModel
+          .findOne({ assemblyId, name: currentSeqId })
+          .exec()
+        if (!refSeqDoc) {
+          throw new NotFoundException(
+            `RefSeq was not found by assemblyId "${assemblyId}" and seq_id "${currentSeqId}" not found`,
+          )
+        }
+        refSeqId = refSeqDoc._id
+        this.logger.debug(
+          `Added new feature for refSeq "${refSeqId}" into database`,
+        )
+        await this.featureModel.create({
+          refSeqId,
+          featureId: featureIdArray,
+          gff3FeatureLineWithRefs: entry,
+        })
+        cnt++
+      }
+    }
+    this.logger.verbose(`Added ${cnt} features into database`)
+  }
+
+  /**
+   * Loop child features in parent feature and add featureId to each child's attribute
+   * @param parentFeature - Parent feature
+   */
+  setAndGetFeatureIdRecursively(
+    parentFeature: GFF3FeatureLineWithRefsAndFeatureId,
+    featureIdArrAsParam: string[],
+  ): string[] {
+    this.logger.verbose(
+      `Value in recursive method = ${JSON.stringify(parentFeature)}`,
+    )
+    // If there is child features and size is not 0
+    if (
+      parentFeature.hasOwnProperty('child_features') &&
+      Object.keys(parentFeature.child_features).length > 0
+    ) {
+      // Loop each child feature
+      for (
+        let i = 0;
+        i < Object.keys(parentFeature.child_features).length;
+        i++
+      ) {
+        // There can be several features with same ID so we need to loop
+        for (let j = 0; parentFeature.child_features[i].length > j; j++) {
+          const assignedVal: GFF3FeatureLineWithRefsAndFeatureId =
+            Object.assign(parentFeature.child_features[i][j])
+          // Let's add featureId if it doesn't exist yet
+          if (!assignedVal.hasOwnProperty('featureId')) {
+            const uid = uuidv4()
+            assignedVal.featureId = uid
+            this.logger.verbose(
+              `FeatureId assigned (recursive level) ${JSON.stringify(
+                assignedVal,
+              )}`,
+            )
+            featureIdArrAsParam.push(uid)
+          }
+          this.setAndGetFeatureIdRecursively(assignedVal, featureIdArrAsParam)
+        }
+      }
```

Comment:
This can be a lot more concise with array mapping. Also, no need to check for `featureId` since we don't support loading those from GFF3.

```suggestion
    // If there are child features
    if (parentFeature.child_features) {
      parentFeature.child_features = parentFeature.child_features.map(
        (childFeature) =>
          childFeature.map((childFeatureLine) => {
            const featureId = uuidv4()
            featureIdArrAsParam.push(featureId)
            const newChildFeature = { ...childFeatureLine, featureId }
            this.setAndGetFeatureIdRecursively(
              newChildFeature,
              featureIdArrAsParam,
            )
            return newChildFeature
          }),
      ) as GFF3FeatureLineWithRefsAndFeatureId[][]
```

Reply:
Fixed as proposed.